### PR TITLE
Hessian-based score w.r.t weights (Pytorch)

### PR DESCRIPTION
**Pull request by eladc-git**
Add hessian-based scores calculation w.r.t weights (Pytorch Implementation)

### CODE REVIEW

File: model_compression_toolkit/core/pytorch/hessian/activation_trace_hessian_calculator_pytorch.py [hunk 1]
```diff
@@ -76,7 +75,7 @@ def compute(self) -> List[float]:
 
             # Concat outputs
             # First, we need to unfold all outputs that are given as list, to extract the actual output tensors
-            output = self._concat_tensors(output_tensors)
+            output = self.concat_tensors(output_tensors)
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/hessian/activation_trace_hessian_calculator_pytorch.py:78]: Same question as 'unfold' question

File: model_compression_toolkit/core/pytorch/hessian/weights_trace_hessian_calculator_pytorch.py [hunk 2]
```diff
@@ -14,37 +14,109 @@
...
                                                                    input_images=input_images,
                                                                    fw_impl=fw_impl,
-                                                                   trace_hessian_request=trace_hessian_request)
+                                                                   trace_hessian_request=trace_hessian_request,
+                                                                   num_iterations_for_approximation=num_iterations_for_approximation)
+
+
+    def compute(self) -> np.ndarray:
+        """
+        Compute the Hessian-based scores w.r.t target node's weights.
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/hessian/weights_trace_hessian_calculator_pytorch.py:59]: Please add to the docs some examples of returned shapes based on the requested granularity.

File: model_compression_toolkit/core/pytorch/hessian/weights_trace_hessian_calculator_pytorch.py [hunk 3]
```diff
@@ -14,37 +14,109 @@
...
+                                     inputs=weights_tensor,
+                                     retain_graph=True)[0]
+
+            # Trace{A^T * A} = sum of all squares values of A
+            approx = f_v_grad ** 2
+            if len(shape_channel_axis) > 0:
+                approx = torch.sum(approx, dim=shape_channel_axis)
+
+            if j > MIN_JACOBIANS_ITER:
+                new_mean = (torch.sum(torch.stack(approximation_per_iteration))+approx)/(j+1)
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/hessian/weights_trace_hessian_calculator_pytorch.py:110]: Are you sure it computes the mean in the per-channel and per-element case? If it's I would add the axis to the sum just to make it clearer

File: tests/pytorch_tests/function_tests/test_hessian_info_weights.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
+# Copyright 2022 Sony Semiconductor Israel, Inc. All rights reserved.
```
Comment (reuvenperetz) [inline tests/pytorch_tests/function_tests/test_hessian_info_weights.py:1]: Fix license year

File: tests/pytorch_tests/function_tests/test_hessian_info_weights.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        t.requires_grad_()
+        inputs.append(t)
+    inputs = to_torch_tensor(inputs)
+    return inputs
+
+
+def get_expected_shape(weights_shape, granularity):
+    if granularity==hessian_common.HessianInfoGranularity.PER_ELEMENT:
+        return weights_shape
+    elif granularity==hessian_common.HessianInfoGranularity.PER_TENSOR:
```
Comment (reuvenperetz) [inline tests/pytorch_tests/function_tests/test_hessian_info_weights.py:105]: I think it's better to output a clear shape like (1,)

File: tests/pytorch_tests/function_tests/test_hessian_info_weights.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+                                              interest_point=ipt,
+                                              granularity=hessian_common.HessianInfoGranularity.PER_ELEMENT)
+
+
+class WeightsHessianTraceAdvanceModelTest(BasePytorchTest):
+    def __init__(self, unit_test):
+        super().__init__(unit_test)
+        self.val_batch_size = 4
+
+    def create_inputs_shape(self):
```
Comment (reuvenperetz) [inline tests/pytorch_tests/function_tests/test_hessian_info_weights.py:169]: Use smaller input shapes when possible

File: tests/pytorch_tests/function_tests/test_hessian_info_weights.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+                                              granularity=hessian_common.HessianInfoGranularity.PER_TENSOR)
+            test_weights_hessian_trace_approx(hessian_service,
+                                              interest_point=ipt,
+                                              granularity=hessian_common.HessianInfoGranularity.PER_ELEMENT)
+
+
+class WeightsHessianTraceAdvanceModelTest(BasePytorchTest):
+    def __init__(self, unit_test):
+        super().__init__(unit_test)
+        self.val_batch_size = 4
```
Comment (reuvenperetz) [inline tests/pytorch_tests/function_tests/test_hessian_info_weights.py:167]: Why does it need a batch size of 4?

File: tests/pytorch_tests/function_tests/test_function_runner.py [hunk 8]
```diff
@@ -112,6 +114,14 @@ def test_model_gradients(self):
         ModelGradientsNonDifferentiableNodeModelTest(self).run_test()
         ModelGradientsSinglePointTest(self).run_test()
 
+    def test_weights_hessian_trace(self):
+        """
```
Comment (reuvenperetz) [inline tests/pytorch_tests/function_tests/test_function_runner.py:118]: Make sure it runs in the test suite

File: tests/pytorch_tests/function_tests/test_hessian_info_weights.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+class multiple_outputs_model(torch.nn.Module):
+    def __init__(self):
+        super(multiple_outputs_model, self).__init__()
+        self.conv1 = Conv2d(3, 3, kernel_size=1, stride=1)
+        self.bn1 = BatchNorm2d(3)
+        self.relu1 = ReLU()
+        self.conv2 = Conv2d(3, 3, kernel_size=1, stride=1)
+        self.bn2 = BatchNorm2d(3)
+        self.hswish = Hardswish()
+        self.dense = Linear(32, 7)
```
Comment (reuvenperetz) [inline tests/pytorch_tests/function_tests/test_hessian_info_weights.py:79]: We can use smaller model for the test purposes

File: tests/pytorch_tests/function_tests/test_hessian_info_weights.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+                                              granularity=hessian_common.HessianInfoGranularity.PER_ELEMENT)
+
+
+class WeightsHessianTraceMultipleOutputsModelTest(BasePytorchTest):
+    def __init__(self, unit_test):
+        super().__init__(unit_test)
+        self.val_batch_size = 1
+
+    def create_inputs_shape(self):
+        return [[self.val_batch_size, 3, 32, 32]]
```
Comment (reuvenperetz) [inline tests/pytorch_tests/function_tests/test_hessian_info_weights.py:210]: Same comment (for all tests): Please use the smallest needed input shape and models for unit tests.

File: tests/pytorch_tests/function_tests/test_function_runner.py [hunk 11]
```diff
@@ -112,6 +114,14 @@ def test_model_gradients(self):
...
+        This test checks the weighes hessian trace approximation in Pytorch.
+        """
+        WeightsHessianTraceBasicModelTest(self).run_test()
+        WeightsHessianTraceAdvanceModelTest(self).run_test()
+        WeightsHessianTraceMultipleOutputsModelTest(self).run_test()
```
Comment (reuvenperetz) [inline tests/pytorch_tests/function_tests/test_function_runner.py:123]: I would also check:
A model with multiple inputs.
A model with a node that has multiple outputs (like split) which is not the model's output.
A model with a reused layer.

File: model_compression_toolkit/core/pytorch/hessian/weights_trace_hessian_calculator_pytorch.py (hunk 3 above)
Comment (eladc-git) [reply model_compression_toolkit/core/pytorch/hessian/weights_trace_hessian_calculator_pytorch.py:110]: you are right. I forgot the axis.

File: model_compression_toolkit/core/common/hessian/trace_hessian_calculator.py [hunk 12]
```diff
@@ -76,16 +76,14 @@ def compute(self) -> List[float]:
         """
         raise NotImplemented(f'{self.__class__.__name__} have to implement compute method.')  # pragma: no cover
 
-    def _unfold_tensors_list(self, tensors_to_unfold: Any) -> List[Any]:
+    @staticmethod
+    def unfold_tensors_list(tensors_to_unfold: Any) -> List[Any]:
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/core/common/hessian/trace_hessian_calculator.py:80]: "A single leading underscore in front of a variable, a function, or a method name means that these objects are used internally". Definitely, here it is not internal cause we use it in other classes.

File: tests/pytorch_tests/function_tests/test_hessian_info_weights.py (hunk 7 above)
Comment (eladc-git) [reply tests/pytorch_tests/function_tests/test_hessian_info_weights.py:167]: I will change it to 2. (to test batch size > 1)

File: tests/pytorch_tests/function_tests/test_hessian_info_weights.py (hunk 5 above)
Comment (eladc-git) [reply tests/pytorch_tests/function_tests/test_hessian_info_weights.py:105]: numpy array of scalar is with shape=()

File: tests/pytorch_tests/function_tests/test_function_runner.py (hunk 8 above)
Comment (eladc-git) [reply tests/pytorch_tests/function_tests/test_function_runner.py:118]: it is: part of "test_function_runner"

File: tests/pytorch_tests/function_tests/test_hessian_info_weights.py (hunk 5 above)
Comment (reuvenperetz) [reply tests/pytorch_tests/function_tests/test_hessian_info_weights.py:105]: I know. I think it's better to align it with what Keras returns. And in general it will makes the tests and code clearer

**Review by reuvenperetz — APPROVED** [review]
